design-proposals/klip-66-multi-param-udafs.md

```diff
@@ -0,0 +1,165 @@
+# KLIP 66 - Multiple parameter UDAFs
+
+**Author**: @jzaralim | 
+**Release Target**: TBD | 
+**Status**: In Discussion | 
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** _Enable users to create UDAFs with multiple parameters._
+
+## Motivation and background
+
+Currently, UDAFs can only accept a single column to aggregate. Users have shown interest
+([#5747](https://github.com/confluentinc/ksql/issues/5747), [#7435](https://github.com/confluentinc/ksql/issues/7435), [#5300](https://github.com/confluentinc/ksql/issues/5300))
+in aggregating over multiple columns. The workaround is to combine all of the required columns into a `Struct` and then implementing a `UDAF` to
+work with `Struct` with that specific schema. This is tedious, and requires the user to implement a new version of their function for every schema they need to work with.
+
+## What is in scope
+
+* UDAFs that accept variadic parameters: `example(col_int_1, col_int_2, col_int_3...) -> result`
+* UDAFs that accept a specific number of parameters, potentially of different types: `example(col_int, col_string) -> result`
+* Variadic TopK - a version of `TOPK` that returns the top K values of one column, along with the values of other columns from those records.
+* `correlation(double, double)` - a UDAF that computes the correlation coefficient of two columns.
+
+## What is not in scope
+
+* UDAFs that return multiple columns. 
+
+## Design
+
+### Variadic parameters
+
+We will add a new class called `MultipleArgs` for users to represent variadic arguments, which is a wrapper for `List`:
+
+```
+public class MultipleArgs<T> {
+
+  private final List<T> values;
+
+  public MultipleArgs(final List<T> values) {
+    this.values = values;
+  }
+
+  public List<T> getValues() {
+    return values;
+  }
+}
+```
+
+The `List` contains the values of the columns specified in the function call. An example usage is:
+```
+// Returns the sum of all the values from multiple colunms
+private static <Integer> Udaf<MultipleArgs<Integer>, Long, Long> MultiColumnSum() {
+  return new Udaf<MultipleArgs<Integer>, List<Integer>, Long>() {
```

Comment:
I think you've got a type mismatch for the aggregate types `Long` and `List<Integer>`. 

Silly Markdown doesn't come up with code highlighting;).